### PR TITLE
Co-Z + effective affine precomputation + tests

### PR DESCRIPTION
**Pull request by sipa**
This builds on top of #171 but reworks the group element tests, making them much more extensive (and faster).

### CODE REVIEW

**Comment by sipa** [comment]
Rebased, and restructured the commits significantly (moving the Co-Z part to a separate commit). @peterdettman since the code and ideas were yours, I left you as author, but feel free to complain if you think that's not appropriate.

**Comment by peterdettman** [comment]
@sipa I don't have any problems with that, thanks for asking though.

**Comment by gmaxwell** [comment]
@sipa thanks for rebasing

**Comment by sipa** [comment]
Rebased again.

**Comment by gmaxwell** [comment]
Needs operation counts.

**Comment by sipa** [comment]
Done.

File: configure.ac [hunk 1]
```diff
@@ -305,11 +310,16 @@ if test x"$use_endomorphism" = x"yes"; then
   AC_DEFINE(USE_ENDOMORPHISM, 1, [Define this symbol to use endomorphism optimization])
 fi
 
+if test x"$use_coz" = x"yes"; then
+  AC_DEFINE(USE_COZ, 1, [Define this symbol to use endomorphism optimization])
```
Comment (peterdettman) [inline configure.ac]: Copy/paste error for help text here?

Comment (sipa) [inline configure.ac]: > Copy/paste error for help text here?

Fixed.

**Comment by sipa** [comment]
Refactored the code further (to facilitate later batch chaining, in order to support batch signature validation).

**Comment by peterdettman** [comment]
@sipa I put together a commit demonstrating how to get half of the speed gain of Co-Z precomp without the Co-Z formulae: https://github.com/peterdettman/secp256k1/commit/f188650d2a54434907ec1f655aa60ca33ac287e8 . It basically applies the effective-affine trick so that the precomp additions are mixed.

**Comment by sipa** [comment]
@peterdettman Awesome, makes perfect sense.

Benchmarks (without GLV, with GMP):
Effective affine: 256k cycles
Effective affine + your endomorphism odd multiples: 254k cycles
Effective affine + Co-Z odd multiples: 248k cycles

Benchmarks (with GLV, with GMP):
Effective affine: 188k cycles
Effictive affine + your endomorphism odd multiples: 185k cycles
Effective affine + Co-Z add multiples: 182k cycles

**Comment by sipa** [comment]
Closing this, way too many refactors to follow the discussion; I'll open new pull requests.

**Pull request closed without merging**